### ISSUE TITLE
StackOverflowError when running tests with higher load

### ISSUE THREAD

**Issue by magro** [gatling/gatling#3254]
When we're running our load tests with increased load the test fails with a StackOverflowError. Because our perf test setup is a bit complicated I cannot easily provide a reproducable sample. At first I wanted to share the stack trace here, maybe this already helps to see the underlying issue or s.th. that could be changed in gatling internals.

```
build	02-Mar-2017 18:24:37	stderr: Uncaught error from thread [GatlingSystem-akka.actor.default-dispatcher-14] shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[GatlingSystem]
build	02-Mar-2017 18:24:37	java.lang.StackOverflowError
build	02-Mar-2017 18:24:37		at scala.runtime.ScalaRunTime$._toString(ScalaRunTime.scala:166)
build	02-Mar-2017 18:24:37		at com.company.project.domain.common.SomeEntity.toString(SomeEntity.scala:105)
build	02-Mar-2017 18:24:37		at java.lang.String.valueOf(String.java:2994)
build	02-Mar-2017 18:24:37		at scala.collection.mutable.StringBuilder.append(StringBuilder.scala:200)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$$anonfun$addString$1.apply(TraversableOnce.scala:359)
build	02-Mar-2017 18:24:37		at scala.collection.Iterator$class.foreach(Iterator.scala:893)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$class.addString(TraversableOnce.scala:357)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.addString(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$class.mkString(TraversableOnce.scala:323)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.mkString(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.runtime.ScalaRunTime$._toString(ScalaRunTime.scala:166)
build	02-Mar-2017 18:24:37		at com.company.project.performance.UseCase1Scenario$Input$3.toString(UseCaseScenario.scala:31)
build	02-Mar-2017 18:24:37		at java.lang.String.valueOf(String.java:2994)
build	02-Mar-2017 18:24:37		at scala.collection.mutable.StringBuilder.append(StringBuilder.scala:200)
build	02-Mar-2017 18:24:37		at scala.collection.MapLike$$anonfun$addString$1.apply(MapLike.scala:344)
build	02-Mar-2017 18:24:37		at scala.collection.MapLike$$anonfun$addString$1.apply(MapLike.scala:344)
build	02-Mar-2017 18:24:37		at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
build	02-Mar-2017 18:24:37		at scala.collection.Iterator$class.foreach(Iterator.scala:893)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$class.addString(TraversableOnce.scala:357)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.addString(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.collection.MapLike$class.addString(MapLike.scala:344)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractMap.addString(Map.scala:59)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$class.mkString(TraversableOnce.scala:323)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractTraversable.mkString(Traversable.scala:104)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableLike$class.toString(TraversableLike.scala:600)
build	02-Mar-2017 18:24:37		at scala.collection.MapLike$class.toString(MapLike.scala:353)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractMap.toString(Map.scala:59)
build	02-Mar-2017 18:24:37		at java.lang.String.valueOf(String.java:2994)
build	02-Mar-2017 18:24:37		at scala.collection.mutable.StringBuilder.append(StringBuilder.scala:200)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$$anonfun$addString$1.apply(TraversableOnce.scala:364)
build	02-Mar-2017 18:24:37		at scala.collection.Iterator$class.foreach(Iterator.scala:893)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$class.addString(TraversableOnce.scala:357)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.addString(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.collection.TraversableOnce$class.mkString(TraversableOnce.scala:323)
build	02-Mar-2017 18:24:37		at scala.collection.AbstractIterator.mkString(Iterator.scala:1336)
build	02-Mar-2017 18:24:37		at scala.runtime.ScalaRunTime$._toString(ScalaRunTime.scala:166)
build	02-Mar-2017 18:24:37		at io.gatling.core.session.Session.toString(Session.scala:71)
build	02-Mar-2017 18:24:37		at java.lang.String.valueOf(String.java:2994)
build	02-Mar-2017 18:24:37		at java.lang.StringBuilder.append(StringBuilder.java:131)
build	02-Mar-2017 18:24:37		at scala.StringContext.standardInterpolator(StringContext.scala:125)
build	02-Mar-2017 18:24:37		at scala.StringContext.s(StringContext.scala:95)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.$bang(Action.scala:63)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.Pause.io$gatling$core$action$ExitableAction$$super$$bang(Pause.scala:34)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.io$gatling$core$action$ExitableAction$class$$$anonfun$3(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.io$gatling$core$action$ExitableAction$class$$$anonfun$3$adapted(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$.exitOrElse(BlockExit.scala:125)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.$bang(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.Pause.$bang(Pause.scala:34)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.InnerTryMax.execute(TryMax.scala:67)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.BlockExit.exitBlock(BlockExit.scala:37)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$.exitOrElse(BlockExit.scala:126)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.$bang(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.builder.SessionHookBuilder$$anon$1.$bang(SessionHookBuilder.scala:35)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.io$gatling$core$action$If$$$anonfun$1(If.scala:44)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.io$gatling$core$action$If$$$anonfun$1$adapted(If.scala:42)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Success.map(Validation.scala:32)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Validation$class.map$mcZ$sp(Validation.scala:19)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Success.map$mcZ$sp(Validation.scala:31)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.execute(If.scala:42)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.io$gatling$core$action$ChainableAction$class$$$anonfun$1(Action.scala:70)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.io$gatling$core$action$ChainableAction$class$$$anonfun$1$adapted(Action.scala:68)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Failure.onFailure(Validation.scala:48)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.recover(Action.scala:68)
build	02-Mar-2017 18:24:37		at io.gatling.http.action.RequestAction.recover(RequestAction.scala:23)
build	02-Mar-2017 18:24:37		at io.gatling.http.action.RequestAction.execute(RequestAction.scala:29)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.$bang(Action.scala:64)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.Pause.io$gatling$core$action$ExitableAction$$super$$bang(Pause.scala:34)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.io$gatling$core$action$ExitableAction$class$$$anonfun$3(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.io$gatling$core$action$ExitableAction$class$$$anonfun$3$adapted(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$.exitOrElse(BlockExit.scala:125)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.$bang(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.Pause.$bang(Pause.scala:34)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.InnerTryMax.execute(TryMax.scala:67)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.BlockExit.exitBlock(BlockExit.scala:37)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$.exitOrElse(BlockExit.scala:126)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.$bang(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.builder.SessionHookBuilder$$anon$1.$bang(SessionHookBuilder.scala:35)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.io$gatling$core$action$If$$$anonfun$1(If.scala:44)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.io$gatling$core$action$If$$$anonfun$1$adapted(If.scala:42)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Success.map(Validation.scala:32)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Validation$class.map$mcZ$sp(Validation.scala:19)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Success.map$mcZ$sp(Validation.scala:31)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.execute(If.scala:42)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.io$gatling$core$action$ChainableAction$class$$$anonfun$1(Action.scala:70)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.io$gatling$core$action$ChainableAction$class$$$anonfun$1$adapted(Action.scala:68)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Failure.onFailure(Validation.scala:48)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.recover(Action.scala:68)
build	02-Mar-2017 18:24:37		at io.gatling.http.action.RequestAction.recover(RequestAction.scala:23)
build	02-Mar-2017 18:24:37		at io.gatling.http.action.RequestAction.execute(RequestAction.scala:29)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.$bang(Action.scala:64)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.Pause.io$gatling$core$action$ExitableAction$$super$$bang(Pause.scala:34)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.io$gatling$core$action$ExitableAction$class$$$anonfun$3(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.io$gatling$core$action$ExitableAction$class$$$anonfun$3$adapted(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$.exitOrElse(BlockExit.scala:125)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.$bang(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.Pause.$bang(Pause.scala:34)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.InnerTryMax.execute(TryMax.scala:67)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.BlockExit.exitBlock(BlockExit.scala:37)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$.exitOrElse(BlockExit.scala:126)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ExitableAction$class.$bang(BlockExit.scala:140)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.builder.SessionHookBuilder$$anon$1.$bang(SessionHookBuilder.scala:35)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.io$gatling$core$action$If$$$anonfun$1(If.scala:44)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.io$gatling$core$action$If$$$anonfun$1$adapted(If.scala:42)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Success.map(Validation.scala:32)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Validation$class.map$mcZ$sp(Validation.scala:19)
build	02-Mar-2017 18:24:37		at io.gatling.commons.validation.Success.map$mcZ$sp(Validation.scala:31)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.If.execute(If.scala:42)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.io$gatling$core$action$ChainableAction$class$$$anonfun$1(Action.scala:70)
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.io$gatling$core$action$ChainableAction$class$$$anonfun$1$adapted(Action.scala:68)
...
```

Nevertheless, here's an excerpt of how scenarios are set up:
```scala

  private def scenarios(loadShare: Double, scenarioFactory: (String, ScenarioContext) => ScenarioBuilder): List[PopulationBuilder] =
    populateScenarios(loadShare, scenarioFactory(_, context))

  private def populateScenarios(loadShare: Double, scenarioFactory: String => ScenarioBuilder): List[PopulationBuilder] =
    protocols.map { protocol =>
      val usersPerSec = loadFactor / loadShare
      scenarioFactory(protocol.baseUrls.head)
        .inject(constantUsersPerSec(usersPerSec.max(1d / simulationTime.toSeconds)) during simulationTime)
        .protocols(protocol)
    }

  val useCase1Scenarios = scenarios(4, UseCase1Scenario(baseSourceFiles))
  val useCase2Scenarios = scenarios(4, UseCase2Scenario(baseSourceFiles))
  ...
  setUp(
    useCase1Scenarios ::: useCase2Scenarios ::: ...
  ).assertions(...)

  // as an example the UseCase1Scenario looks like:
  object UseCase1Scenario {
    def apply(...): ScenarioBuilder = {

    ...

    scenario(s"UseCase1Scenario-$id")
      .feed(toFeeder(inputGen))
      .exec(useCase1Request(input(_).sourceFile, input(_).destinationFile, input(_).profile))
      .pause(5)
      // use case 1 might take longer than other tasks
      .poll(pollMax * 2)(pollJobIsComplete())
      .exitHereIfFailed
      .poll(pollMax)(pollFileExists(input(_).destinationFile))
      .exitHereIfFailed
      .pause(1)
      // clean up
      .exec(deleteRequest(input(_).destinationFile))
  }

```

Of course there are many missing pieces, but perhaps this gives an idea how we create our scenarios. Again I'm sorry that I don't have a reproducer, but wanted to share this anyways.

**Comment by slandelle** [comment]
Hey @magro 

Thanks for reporting. Which Gatling version do you run?

**Comment by magro** [comment]
I had updated to 2.2.3 to see if this is happening with the latest version, so 2.2.3.

**Comment by slandelle** [comment]
Weird thing is that line numbers don't match:
```
build	02-Mar-2017 18:24:37		at io.gatling.core.action.ChainableAction$class.$bang(Action.scala:63)
```

vs https://github.com/gatling/gatling/blob/2.2/gatling-core/src/main/scala/io/gatling/core/action/Action.scala#L63

Do you run a fork?

Assuming the proper line is actually https://github.com/gatling/gatling/blob/2.2/gatling-core/src/main/scala/io/gatling/core/action/Action.scala#L64, `Session` should only be serialized if you have INFO logging enabled. Can you confirm?

**Comment by magro** [comment]
Ouch, good catch. I had reported observations from our CI environment. Last night I took the time to setup the stuff locally, there I could run with load 8 and did not get a SOE (I couldn't try more load due to overload). With your observation this sounds as if our CI environment is not cleaning things properly and still has the former gatling version around. I'm just heading to vacations, will check with my colleagues. Many thanks already!
